[GPUP] Chinese text in language selector in Google News is invisible

litherum opened the pull request:
#### 457dbbe185b3f5d84132411a86cb4f368189312c
<pre>
[GPUP] Chinese text in language selector in Google News is invisible
<a href="https://bugs.webkit.org/show_bug.cgi?id=242674">https://bugs.webkit.org/show_bug.cgi?id=242674</a>
&lt;rdar://problem/96903146&gt;

Reviewed by Cameron McCormack.

When falling off the end of the font-family list, the browser needs to pick some font which
can render the requested character. Font selection routines accept a &quot;hint&quot; font, which it
tries to match as stylistically closely as possible. The hint font itself may actually be
selected - if it is, we have to make sure the fallback font takes the CreationData from the
hint font. However, if the hint font is not selected, then the fallback font will be
guaranteed to be an installed font, which means that no CreationData should be specified.
252096@main had some of this logic, but not all of it.

* LayoutTests/fast/text/fallback-not-supported-expected-mismatch.html: Added.
* LayoutTests/fast/text/fallback-not-supported.html: Added.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::systemFallbackForCharacters):

Canonical link: <a href="https://commits.webkit.org/252407@main">https://commits.webkit.org/252407@main</a>
</pre>

heycam commented on Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
```diff
@@ -1565,7 +1565,10 @@ RefPtr<Font> FontCache::systemFallbackForCharacters(const FontDescription& descr
 
-    FontPlatformData alternateFont(substituteFont, platformData.size(), syntheticBold, syntheticOblique, platformData.orientation(), platformData.widthVariant(), platformData.textRenderingMode(), platformData.creationData());
+    const FontPlatformData::CreationData* creationData = nullptr;
+    if (safeCFEqual(platformData.font(), substituteFont))
+        creationData = platformData.creationData();
```
Can we assert in here that if `platformData` is not `substituteFont`, then it's an installed font?

litherum replied:
Hrmmm I don't see a way to do that 🤔

litherum replied:
I'll land as-is and in parallel discuss with the Core Text team.

litherum committed:
[GPUP] Chinese text in language selector in Google News is invisible
https://bugs.webkit.org/show_bug.cgi?id=242674
<rdar://problem/96903146>

Reviewed by Cameron McCormack.

When falling off the end of the font-family list, the browser needs to pick some font which
can render the requested character. Font selection routines accept a "hint" font, which it
tries to match as stylistically closely as possible. The hint font itself may actually be
selected - if it is, we have to make sure the fallback font takes the CreationData from the
hint font. However, if the hint font is not selected, then the fallback font will be
guaranteed to be an installed font, which means that no CreationData should be specified.
252096@main had some of this logic, but not all of it.

* LayoutTests/fast/text/fallback-not-supported-expected-mismatch.html: Added.
* LayoutTests/fast/text/fallback-not-supported.html: Added.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::systemFallbackForCharacters):

Canonical link: https://commits.webkit.org/252407@main
